File: scripts/check_vocab_coverage.py
import json
import sys
from collections import Counter

def main(vocab_path, data_path):
    with open(vocab_path) as f:
        vocab = json.load(f)
    with open(data_path) as f:
        data = json.load(f)
    vocab_set = set(vocab.keys())
    missing = Counter()
    for entry in data:
        if 'semantic_output' in entry:
            def recurse(obj):
                if isinstance(obj, dict):
                    for k, v in obj.items():
                        if k not in vocab_set:
                            missing[k] += 1
                        recurse(v)
                elif isinstance(obj, list):
                    for item in obj:
                        recurse(item)
                else:
                    if str(obj) not in vocab_set:
                        missing[str(obj)] += 1
            recurse(entry['semantic_output'])
    print("Missing tokens:")
    for token, count in missing.most_common():
        print(f"{token}: {count}")

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: python check_vocab_coverage.py <vocab.json> <data.json>")
        sys.exit(1)
    main(sys.argv[1], sys.argv[2])

File: scripts/check_vocab_coverage.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"
	"sort"
)

// Pair is a struct to hold key-value pairs for sorting.
type Pair struct {
	Key   string
	Value int
}

// PairList is a slice of Pairs that implements sort.Interface.
type PairList []Pair

func (p PairList) Len() int           { return len(p) }
func (p PairList) Less(i, j int) bool { return p[i].Value < p[j].Value }
func (p PairList) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }

func findMissing(obj interface{}, vocab map[string]int, missing map[string]int) {
	switch v := obj.(type) {
	case map[string]interface{}:
		for k, val := range v {
			if _, exists := vocab[k]; !exists {
				missing[k]++
			}
			findMissing(val, vocab, missing)
		}
	case []interface{}:
		for _, item := range v {
			findMissing(item, vocab, missing)
		}
	default:
		token := fmt.Sprintf("%v", v)
		if _, exists := vocab[token]; !exists {
			missing[token]++
		}
	}
}

func main() {
	if len(os.Args) != 3 {
		fmt.Println("Usage: go run check_vocab_coverage.go <vocab.json> <data.json>")
		os.Exit(1)
	}
	vocabPath := os.Args[1]
	dataPath := os.Args[2]

	// Read and parse the vocabulary file
	vocabFile, err := ioutil.ReadFile(vocabPath)
	if err != nil {
		fmt.Printf("Error reading vocab file: %v\n", err)
		os.Exit(1)
	}

	var vocab map[string]int
	if err := json.Unmarshal(vocabFile, &vocab); err != nil {
		fmt.Printf("Error unmarshalling vocab json: %v\n", err)
		os.Exit(1)
	}

	// Read and parse the data file
	dataFile, err := ioutil.ReadFile(dataPath)
	if err != nil {
		fmt.Printf("Error reading data file: %v\n", err)
		os.Exit(1)
	}

	var data []map[string]interface{}
	if err := json.Unmarshal(dataFile, &data); err != nil {
		fmt.Printf("Error unmarshalling data json: %v\n", err)
		os.Exit(1)
	}

	missing := make(map[string]int)
	for _, entry := range data {
		if semanticOutput, ok := entry["semantic_output"]; ok {
			findMissing(semanticOutput, vocab, missing)
		}
	}

	// Create a slice of pairs to sort the map by value.
	p := make(PairList, len(missing))
	i := 0
	for k, v := range missing {
		p[i] = Pair{k, v}
		i++
	}

	// Sort the slice by value in descending order.
	sort.Sort(sort.Reverse(p))

	fmt.Println("Missing tokens:")
	for _, k := range p {
		fmt.Printf("%s: %d\n", k.Key, k.Value)
	}
}
